pkg/cvo/upgradeable.go

```diff
@@ -234,6 +247,131 @@ func (check *clusterManifestDeleteInProgressUpgradeable) Check() *configv1.Clust
+	cvLister         configlistersv1.ClusterVersionLister
+	cvoName          string
+}
+
+func (check *clusterAdminAcksCompletedUpgradeable) Check() *configv1.ClusterOperatorStatusCondition {
+	cv, err := check.cvLister.Get(check.cvoName)
+	if meta.IsNoMatchError(err) || apierrors.IsNotFound(err) {
+		message := fmt.Sprintf("Unable to get ClusterVersion, err=%v.", err)
+		klog.Error(message)
+		return &configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.ClusterStatusConditionType("UpgradeableAdminAckRequired"),
+			Status:  configv1.ConditionFalse,
+			Reason:  "UnableToGetClusterVersion",
+			Message: message,
+		}
+	}
+	currentVersion := clusterversion.GetCurrentVersion(cv.Status.History)
+
+	// This can occur in early start up when the configmap is first added and version history
+	// has not yet been populated.
+	if currentVersion == "" {
+		return nil
+	}
+
+	var gateCm *corev1.ConfigMap
+	if gateCm, err = check.adminGatesLister.Get(internal.AdminGatesConfigMap); err != nil {
+		var message string
+		if apierrors.IsNotFound(err) {
+			message = fmt.Sprintf("%s configmap not found.", internal.AdminGatesConfigMap)
+		} else if err != nil {
+			message = fmt.Sprintf("Unable to access configmap %s, err=%v.", internal.AdminGatesConfigMap, err)
+		}
+		klog.Error(message)
+		return &configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.ClusterStatusConditionType("UpgradeableAdminAckRequired"),
+			Status:  configv1.ConditionFalse,
+			Reason:  "UnableToAccessAdminGatesConfigMap",
+			Message: message,
+		}
+	}
+	var ackCm *corev1.ConfigMap
+	if ackCm, err = check.adminAcksLister.Get(internal.AdminAcksConfigMap); err != nil {
+		var message string
+		if apierrors.IsNotFound(err) {
+			message = fmt.Sprintf("%s configmap not found.", internal.AdminAcksConfigMap)
+		} else if err != nil {
+			message = fmt.Sprintf("Unable to access configmap %s, err=%v.", internal.AdminAcksConfigMap, err)
+		}
+		klog.Error(message)
+		return &configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.ClusterStatusConditionType("UpgradeableAdminAckRequired"),
+			Status:  configv1.ConditionFalse,
+			Reason:  "UnableToAccessAdminAcksConfigMap",
+			Message: message,
+		}
+	}
+	var reasons []string
+	var messages []string
+	for k, v := range gateCm.Data {
+		if reason, message := checkAdminGate(k, v, currentVersion, ackCm); reason != "" {
+			reasons = append(reasons, reason)
+			messages = append(messages, message)
+		}
+	}
+	if len(reasons) == 1 {
+		return &configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.ClusterStatusConditionType("UpgradeableAdminAckRequired"),
+			Status:  configv1.ConditionFalse,
+			Reason:  reasons[0],
+			Message: messages[0],
+		}
+	} else if len(reasons) > 1 {
+		return &configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.ClusterStatusConditionType("UpgradeableAdminAckRequired"),
+			Status:  configv1.ConditionFalse,
+			Reason:  "MultipleReasons",
```

Comment:
Ideally if we have `len(reasons) > 1` but all the reasons are identical (e.g. `AdminAckRequired`), we'd collapse to a single value here.  Maybe we should collect a map of `reasons`, and then here in the `len > 1` case collapse that to a sorted slice?

```go
exists := struct{}{}
reasons := map[string]struct{}{}
for k, v := range gateCm.Data {
	if reason, message := checkAdminGate(k, v, currentVersion, ackCm); reason != "" {
		reasons[reason] = exists
		messages = append(messages, message)
	}
}
if len(reasons) == 1 {
	condition := ...
	for reason in :=range reasons {
		condition.Reason = reason // or is there a more convenient way to get a key out of a single-member map?
	}
	return condition
} else if len(reasons) > 1 {
	...
}

Reply:
Unique `Reason` collapsing done above and then check for multi-reasons and join/sort messages.